### ISSUE TITLE
Unable to renew expired certificate

### ISSUE THREAD

**Issue by antoinep92** [kubernetes/kubernetes#102556]
#### What happened:
Control plane is broken. Kubelet is technically running but seems stuck in initialization. I think this is a certificate problem, because it stopped working suddenly with no control-plane, hardware or OS-level change, and the following elements:
- `kubectl get nodes` fails with `unauthorized`
- kube-apiserver POD is restarting and complains with:
```
1 authentication.go:63] "Unable to authenticate the request" err="[x509: certificate has expired or is not yet valid: current time 2021-06-03T07:29:14Z is after 2021-05-29T15:48:22Z
```

On the other hand, `kubeadm certs check-expiration` seems happy:
```
CERTIFICATE                EXPIRES                  RESIDUAL TIME   CERTIFICATE AUTHORITY   EXTERNALLY MANAGED
admin.conf                 Jun 01, 2022 19:59 UTC   363d                                    no      
apiserver                  Jun 01, 2022 19:59 UTC   363d            ca                      no      
apiserver-etcd-client      Jun 01, 2022 19:59 UTC   363d            etcd-ca                 no      
apiserver-kubelet-client   Jun 01, 2022 19:59 UTC   363d            ca                      no      
controller-manager.conf    Jun 01, 2022 19:59 UTC   363d                                    no      
etcd-healthcheck-client    Jun 01, 2022 19:59 UTC   363d            etcd-ca                 no      
etcd-peer                  Jun 01, 2022 19:59 UTC   363d            etcd-ca                 no      
etcd-server                Jun 01, 2022 19:59 UTC   363d            etcd-ca                 no      
front-proxy-client         Jun 01, 2022 19:59 UTC   363d            front-proxy-ca          no      
scheduler.conf             Jun 01, 2022 19:59 UTC   363d                                    no      

CERTIFICATE AUTHORITY   EXPIRES                  RESIDUAL TIME   EXTERNALLY MANAGED
ca                      May 27, 2030 15:48 UTC   8y              no      
etcd-ca                 May 27, 2030 15:48 UTC   8y              no      
front-proxy-ca          May 27, 2030 15:48 UTC   8y              no 
```
I tried to renew all certificates anyways and reboot/restart services with no effect.
I also base64-decoded all relevent fields in kubernetes yaml config files and checked with openssl, and found no expired certificate.

I know this is not a support channel, and I'm not even 100% sure this is a bug, and if so if it is in kubeadm or another component. But I have posted this on slack, serverfault, and community forum with no answer, so any tip to how to repair the cluster is welcome :pray: 

#### What you expected to happen:
I expected `kube-apiserver` logs or `kubeadm certs` to tell which certificate is expired. Also, I expected the certificate to be renewed by `kubeadm certs renew all`.

#### How to reproduce it (as minimally and precisely as possible):
Unfortunately I'm not sure :(

#### Anything else we need to know?:
Before finding about this certificate issue, the kubelet was not even starting because of missing `/etc/kubernetes/bootstrap-kubelet.conf` which is a bit mysterious because I really don't think this was deleted manually. I recreated the file but was not sure how to fill the `token` field for user `tls-bootstrap-token-user`. After that `kubelet started again` but as I said, the control plane is still unusable. Not sure if this is related.

#### Environment:
- Kubernetes version (use `kubectl version`): 1.21.0
- Cloud provider or hardware configuration: bare metal
- OS (e.g: `cat /etc/os-release`): Arch Linux
- Kernel (e.g. `uname -a`): 5.11.16
- Install tools: kubeadm
- Network plugin and version (if this is a network-related bug): weave (CNI 0.9.1)
- Others: containerd 1.4.4, CRI 1.21.0

**Comment by neolit123** [comment]
Hi. Kubeadm does not manage the kubelet cert rotation. It is automatic by default unless the user opts out. We have seen cases in the wild where 'something just goes wrong' and a kubelet is stuct without a valid cert.

As long as you have the ca.crt/key on a node you can create a new kubelet certs with:

- Call 'kubeadm init phase kubeconfig kubelet' optionally pass -config or other flags if needed. This will generate kubelet.conf with embedded cert and key.
- Restart kubelet and wait for it to generate new certs in /var/lib/kubelet/pki
- Call 'kubeadm init phase kubelet-finalize all'. This will 'symlink' the kebelet.conf to kubelet cert rotation.

LMK if this works for you. I have explained this so many times time that I think we shpuld add it to the kubeadm cert management page.

**Comment by antoinep92** [comment]
Thanks @neolit123 ! I apologize for the noise if this is a known issue. As you said it is not immediately obvious from the logs/docs what exactly is going on. Anyways regarding your instructions, unfortunately they do not work out of the box. Feedback step by step:
- [no effect] I made a backup of my config files and ran `kubeadm init pahse kubeconfig kubelet -v=5` which wrote `/etc/kubernetes/kubelet.conf` successfully but the new generated file was identical to the previous one (as I said, I played a bit with certs and base64 earlier to try to fix my issue, I guess I got it right).
- [no effect] I already tried restarting kubelet many times, doing it once more did nothing different
- [error] the kubelet-finalize fails with:
```
I0603 15:28:50.074086   28728 kubeletfinalize.go:88] [kubelet-finalize] Assuming that kubelet client certificate rotation is enabled: found "/var/lib/kubelet/pki/kubelet-client-current.pem"
[kubelet-finalize] Updating "/etc/kubernetes/kubelet.conf" to point to a rotatable kubelet client certificate and key
the file "/etc/kubernetes/kubelet.conf" does not contain authentication for user "sc001"
```
`sc001` is the name of my master node. I tried passing options like `--service-cidr 172.20.0.0/16 --apiserver-cert-extra-sans sc001` but they are not meant for this phase apparently. Note that kubelet.conf was still not changed.

But I did find out something interesting: the date of the expired certificate mentioned in the apiserver logs is that of `/var/lib/kubelet/pki/kubelet-client-2020-05-29-17-48-30.pem`. So my guess is the certificate was renewed since kubelet-client-current seems ok, but the change was not applied everywhere, and something is still using the old certificate, unfortunately I don't know where:
```
cd /ect && grep -R $(cat /var/lib/kubelet/pki/kubelet-client-2020-05-29-17-48-30.pem | head -n 18 | base64 -w0)
```
gives no result. But I guess that's what `kubelet-finalize` is for.

**Comment by neolit123** [comment]
> [no effect] I made a backup of my config files and ran kubeadm init pahse kubeconfig kubelet -v=5 which wrote /etc/kubernetes/kubelet.conf successfully but the new generated file was identical to the previous one (as I said, I played a bit with certs and base64 earlier to try to fix my issue, I guess I got it right).

did you try deleting the kubelet.conf first?

`kubeadm init` tries to reuse existing .conf files.

**Comment by neolit123** [comment]
the steps i suggested above work in this online k8s environment:
https://www.katacoda.com/courses/kubernetes/playground

**Comment by antoinep92** [comment]
I didn't kubeadm init with a config file, but it was a few years ago on v1.11
When I delete kubelet.conf the one kubeadm creates uses a completely different file indeed, probably because defaults have changed. I guess kubeadm handled this well each time I upgraded, but not for troubleshooting stuff.

I'll try to recreate a config file that matches my actual cluster config by mapping the various files in /etc/kubernetes, command line flags, etc. to the kubeadm.k8s.io/v1beta2 schema and try again

**Comment by neolit123** [comment]
note that you don't need a --config for `init phase kubeconfig` if:
- you let kubeadm detects the API of the node
technically you can just let it run without config and then amend the kubelet.conf `server: ` field for that.
- kubeadm detects the node hostname, if check check if the kubelet.conf has what you expect:
e.g.
```
- context:
    cluster: kubernetes
    user: system:node:controlplane
  name: system:node:controlplane@kubernetes
```

**Comment by antoinep92** [comment]
Ok many thanks for your help :muscle: 
I was able to `kubectl` again !

I still had to do a few more steps from what you said, but you pointed me in the right direction.

I had to delete /etc/kubelet.conf obviously, and /var/lib/kubelet/pki/kubelet-client-current.pem, which was *not* a symlink but now is. And then I still had to kubeadm certs renew all and update my kubectl config accordingly :ok_hand: 
In the end I'm not quite sure if all the steps were necessary to repair the cluster; but there was definitely something strange with kubelet client certificate.

So again thank you for you time, and frankly I don't know if this qualifies as a bug, but at least I feel the logs and/or documentation could be a bit more helpful ;)
